blockprod/src/detail/timestamp_searcher.rs

```diff
@@ -0,0 +1,487 @@
+// Copyright (c) 2021-2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    sync::{Arc, Mutex},
+};
+
+use rayon::prelude::*;
+
+use chainstate::{
+    chainstate_interface::ChainstateInterface, BlockIndex, ChainstateHandle, NonZeroPoolBalances,
+};
+use chainstate_types::pos_randomness::PoSRandomness;
+use common::{
+    chain::{
+        block::timestamp::BlockTimestamp, config::EpochIndex, ChainConfig, PoSConsensusVersion,
+    },
+    primitives::{BlockHeight, Compact},
+    Uint256,
+};
+use consensus::{
+    calc_pos_hash_from_prv_key, calculate_target_required_from_block_index, check_pos_hash,
+    compact_target_to_target, ConsensusCreationError, ConsensusPoSError,
+    PoSTimestampSearchInputData,
+};
+use crypto::vrf::{VRFPrivateKey, VRFPublicKey};
+use logging::log;
+use randomness::{CryptoRng, Rng};
+use serialization::{Decode, Encode};
+use utils::{ensure, once_destructor::OnceDestructor};
+
+use crate::{
+    detail::utils::{
+        get_best_block_index, get_block_id_from_height, get_existing_block_index,
+        get_existing_gen_block_index, get_pool_balances_at_heights, get_sealed_epoch_randomness,
+        make_ancestor_getter, pos_data_from_header, pos_status_from_height, timestamp_add_secs,
+    },
+    BlockProductionError,
+};
+
+#[derive(Clone, Encode, Decode)]
+struct SearchDataForHeight {
+    sealed_epoch_randomness: PoSRandomness,
+    epoch_index: u64,
+    target_required: Compact,
+    min_timestamp: BlockTimestamp,
+    max_timestamp: BlockTimestamp,
+    pool_balances: NonZeroPoolBalances,
+    consensus_version: PoSConsensusVersion,
+}
+
+#[derive(Clone, Encode, Decode)]
+pub struct TimestampSearchData {
+    start_height: BlockHeight,
+    data: Vec<SearchDataForHeight>,
+    check_all_timestamps_between_blocks: bool,
+}
+
+impl TimestampSearchData {
+    pub fn new(
+        chainstate: &dyn ChainstateInterface,
+        chain_config: &ChainConfig,
+        secret_input_data: &PoSTimestampSearchInputData,
+        min_height: BlockHeight,
+        max_height: Option<BlockHeight>,
+        seconds_to_check_for_height: u64,
+        check_all_timestamps_between_blocks: bool,
+    ) -> Result<Self, BlockProductionError> {
```

Comment:
Perhaps a separate error type for this is a good idea

Reply:
> Perhaps a separate error type for this is a good idea

Actually, initially I tried using a separate `TimestampSearcherError` type for this and I didn't like it.
The main issue I had with this approach is that it makes it harder to re-use utility functions that are common to the "normal" blockprod stuff and timestamp searcher:
1) The utility functions can't return `TimestampSearcherError`, because it's too specific.
2) If they return `BlockProductionError`, then `TimestampSearcherError` has to contain a `BlockProductionError` variant. On the other hand, `BlockProductionError` is an "umbrella" error-type for everything blockprod-related, so `BlockProductionInterface::collect_timestamp_search_data_e2e` should return it as well, so `BlockProductionError` has to contain a `TimestampSearcherError` variant.

Surely, this can be worked around, e.g. by introducing an intermediate error type, or by duplicating some utility functions (which are mostly trivial), but I'd personally leave it as is, unless you insist.

Reply:
It's OK.